### ISSUE TITLE
Equation numbers on aligned/splitted equation numbers do not work properly

### ISSUE THREAD

**Issue by StatProofBook** [StatProofBook/StatProofBook.github.io#7]
When a block of aligned equations appears within a proof, there is supposed to be one equation number that may be used for referencing this block. Unfortunately, there seems to be no configuration which results in this behavior.

When using `\begin{split}...\end{split}`, no equation number is shown. When using `\begin{align}...\end{split}`, every line in the aligned block gets a number. This shouldn't be, because (i) there is just one equation label per such block, so referencing does not work anymore and (ii) sometimes multiple lines do not represent multiple equations, but just multiple parts of a sum or product.

Below comes an example from `Proofs/mvn-ltt.md` and a screenshot of what it generates:

```tex
$$ \label{eq:y-mgf-s1}
\begin{split}
M_y(t) &= \mathbb{E} \left( \exp \left[ t^\mathrm{T} (Ax + b) \right] \right) \\
&= \mathbb{E} \left( \exp \left[ t^\mathrm{T} A x \right] \cdot \exp \left[ t^\mathrm{T} b \right] \right) \\
&= \exp \left[ t^\mathrm{T} b \right] \cdot \mathbb{E} \left( \exp \left[ t^\mathrm{T} A x \right] \right) \\
&= \exp \left[ t^\mathrm{T} b \right] \cdot M_x(At) \; .
\end{split}
$$
```

```tex
$$ \label{eq:y-mgf-s2}
\begin{align}
M_y(t) &= \exp \left[ t^\mathrm{T} b \right] \cdot M_x(At) \\
&= \exp \left[ t^\mathrm{T} b \right] \cdot \exp \left[ t^\mathrm{T} A \mu + \frac{1}{2} t^\mathrm{T} A \Sigma A^\mathrm{T} t \right] \\
&= \exp \left[ t^\mathrm{T} \left( A \mu + b \right) + \frac{1}{2} t^\mathrm{T} A \Sigma A^\mathrm{T} t \right] \; .
\end{align}
$$
```
![Issue](https://user-images.githubusercontent.com/54538843/71258884-8339d400-2337-11ea-8a54-46b03eccc143.png)

How can we solve this problem? Any suggestions appreciated!

**Comment by allefeld** [comment]
If we take the source as LaTeX code, then the second version is actually illegal. `$$`–`$$` defines a math display environment, but `align` is a math display environment itself, defined by the [amsmath](http://www.tug.org/teTeX/tetex-texmfdist/doc/latex/amsmath/amsldoc.pdf) package. `split` on the other hand is an environment *within* math, so it is used inside `$$`–`$$` (or `align`, or `equation`, etc.).

In amsmath, `align` is meant for "Equation groups with mutual alignment". It makes sense that there is more than one number, because there is more than one equation. On the other hand, "The split environment is designed to serve as the entire body of an equation, or an entire line of an align or gather environment.", so it makes sense that it is not numbered by itself – the complete equation or the line is.

So *I guess* the solution would be to use the `split` environment as the entire body of an equation:

```
$$
\begin{equation} \label{eq:y-mgf-s1}
\begin{split}
M_y(t) &= \mathbb{E} \left( \exp \left[ t^\mathrm{T} (Ax + b) \right] \right) \\
&= \mathbb{E} \left( \exp \left[ t^\mathrm{T} A x \right] \cdot \exp \left[ t^\mathrm{T} b \right] \right) \\
&= \exp \left[ t^\mathrm{T} b \right] \cdot \mathbb{E} \left( \exp \left[ t^\mathrm{T} A x \right] \right) \\
&= \exp \left[ t^\mathrm{T} b \right] \cdot M_x(At) \; .
\end{split}
\end{equation}
$$
```

***

*Further ruminations that may or may not be relevant...*

But this is not LaTeX code, but kramdown with embedded LaTeX math. [Here's](https://kramdown.gettalong.org/syntax.html#math-blocks) what kramdown does:

> The content of a math block has to be valid LaTeX math. It is always wrapped inside a \begin{displaymath}...\end{displaymath} enviroment except if it begins with a \begin statement.

But if we look at the HTML generated by kramdown, 

```
<script type="math/tex; mode=display">% <![CDATA[
--
  | \label{eq:y-mgf-s1}
  | \begin{split}
  | M_y(t) &= \mathbb{E} \left( \exp \left[ t^\mathrm{T} (Ax + b) \right] \right) \\
  | &= \mathbb{E} \left( \exp \left[ t^\mathrm{T} A x \right] \cdot \exp \left[ t^\mathrm{T} b \right] \right) \\
  | &= \exp \left[ t^\mathrm{T} b \right] \cdot \mathbb{E} \left( \exp \left[ t^\mathrm{T} A x \right] \right) \\
  | &= \exp \left[ t^\mathrm{T} b \right] \cdot M_x(At) \; .
  | \end{split} %]]></script>
```

```
<script type="math/tex; mode=display">% <![CDATA[
\label{eq:y-mgf-s2}
\begin{align}
M_y(t) &= \exp \left[ t^\mathrm{T} b \right] \cdot M_x(At) \\
&= \exp \left[ t^\mathrm{T} b \right] \cdot \exp \left[ t^\mathrm{T} A \mu + \frac{1}{2} t^\mathrm{T} A \Sigma A^\mathrm{T} t \right] \\
&= \exp \left[ t^\mathrm{T} \left( A \mu + b \right) + \frac{1}{2} t^\mathrm{T} A \Sigma A^\mathrm{T} t \right] \; .
\end{align} %]]></script>
```

we see that there is no explicit `displaymath` environment, even though both begin with `label`, not `\begin`. But maybe `\label` is ignored?! Or is `<script type="math/tex; mode=display">` kramdown's way of saying `\begin{displaymath}`?

Anyway, what does MathJax do with these `script`s? Note that this is not the normal way one would write math in HTML for MathJax, but the way [How mathematics is stored in the page](https://docs.mathjax.org/en/v2.7-latest/advanced/model.html#how-mathematics-is-stored-in-the-page) by MathJax internally. Apparently MathJax' usual first step, to find `\(`–`\)`,  `\[`–`\]`, and `$$`–`$$` in the page text and to convert them into these `script` tags, is already done by kramdown. (However, your inline markup, e.g. "any linear transformation of $x$ is also", is *not* recognized by kramdown – it only recognizes `$$`–`$$` and uses it  for inline math, too – and is therefore only later picked up by MathJax. What a mess!)

Anyway, `<script type="math/tex; mode=display">` represents displayed math within MathJax. Which doesn't actually clear up how one can include both `split` and `align` within these. But MathJax isn't amsmath either, it just emulates it somehow. You use the `TeX-AMS_CHTML` configuration, which includes the AMSmath extension, which enables [a number of math environments](http://docs.mathjax.org/en/v2.7-latest/tex.html#environments), which are not really documented apart from being listed. There is further [user-created documentation](http://www.onemathematicalcat.org/MathJaxDocumentation/TeXSyntax.htm), but it doesn't mention the difference between a displayed math environment and an environment within displayed math either.

If my above guessed solution works, then I would recommend the following:

- Use kramdown's math markup consistently, i.e. put everything in `$$`–`$$`. These look like standard TeX displayed math environments, but they are not really. As far as I understand, kramdown differentiates between displayed and inline math by the context: as a standalone block with empty lines before and after vs inside of text.

- For displayed math, but everything explicitly into an AMS environment, so that the code starts with `\begin{equation}`, `\begin{align}` etc. Put `\label` inside of these environments, and use `\begin{split}` inside of `equation` them if you want to have multiple lines without separate numbers.

**Comment by StatProofBook** [comment]
@allefeld: First of all, many thanks for your effort and time put into solving this issue and for your ruminations additionally enlightening the topic.

To my honest surprise, your guessed solution actually works! I have added `$$ \begin{equation} \label{eq:XYZ} \begin{split} ... \end{split} \end{equation} $$` to [this proof](https://statproofbook.github.io/Proofs/mvn-ltt.html) and there is now exactly one equation number per equation block. I will add this to all proofs currently in the archive and then we can finally move on proving.

Regarding your recommendations at the end, I would object to both – not because I don't like consistency, but because I think it would be good to have maximum freedom for the user (if multiple possibilities work). I will however [document here](https://statproofbook.github.io/Tutorials/Template.html) that there are generally two possibilities for
* _in-line math_, namely either `$...$` or `$$...$$`;
* _single-line equations_, namely either `$$ \label{eq:XYZ} ... $$` or `$$ \begin{equation} \label{eq:XYZ} ... \end{equation} $$`;

and apparently, only one possibility for
* _multi-line equations_, namely `$$ \begin{equation} \label{eq:XYZ} \begin{split} ... \end{split} \end{equation} $$`.

**Comment by StatProofBook** [comment]
@allefeld: OK, it seems we have another problem again! Although splitted equations now have numbers, references to them do no work. In other words, when using `\eqref{eq:XYZ}` where the referenced equation was created via `$$ \begin{equation} \label{eq:XYZ} \begin{split} ... \end{split} \end{equation} $$`, MathJax only displays "(???)".

For example, check `Proofs/glm-mle.md` [where](https://statproofbook.github.io/Proofs/glm-mle.html) references to equations 4, 5, 7 and 9 do not work.

```tex
$$
\begin{equation} \label{eq:GLM-LL1}
\begin{split}
\mathrm{LL}(B,\Sigma) = &\log p(Y|B,\Sigma) \\
= &- \frac{nv}{2} \log(2\pi) - \frac{n}{2} \log |\Sigma| - \frac{v}{2} \log |V| \\
&- \frac{1}{2} \, \mathrm{tr}\left[ \Sigma^{-1} (Y - XB)^\mathrm{T} V^{-1} (Y - XB) \right] \; .\\
\end{split}
\end{equation}
$$
```

```tex
The derivative of the log-likelihood function \eqref{eq:GLM-LL1} with respect to $\Sigma$ is ...
```

Why does all this have to be so complicated?

**Comment by StatProofBook** [comment]
I'm currently using an unreleased [MathJax fix](https://github.com/mathjax/MathJax/issues/2272) (from a fork of the repository) to label and reference single `split` environments and it works well (with and without `equation` environment). Therefore, once this is released MathJax script, I will recommend using `$$ \label{eq:XYZ} \begin{split} ... \end{split} $$` and `\eqref{eq:XYZ}`.

**Comment by allefeld** [comment]
> Why does all this have to be so complicated?

That is a very good question. The way Pandoc handles LaTeX math embedded in Markdown is much more straightforward. Unfortunately, GitHub pages and many other web services based on Markdown only support kramdown or other slightly crappy Markdown implementations. I think it would be possible to use Pandoc, but not automatically, but via a script that has to be run every time something has been changed.

One way to make the situation less messy might be to disable math support in kramdown, and rely on MathJax only.

**Comment by allefeld** [comment]
From https://kramdown.gettalong.org/options.html
> `math_engine`
> Set the math engine
> Specifies the math engine that should be used for converting math blocks/spans. If this option is set to +nil+, no math engine is used and the math blocks/spans are output as is.

Then we would have to embed MathJax manually instead of relying on kramdown to do that, but that shouldn't be too hard.

**Comment by StatProofBook** [comment]
> `math_engine`
> Set the math engine
> Specifies the math engine that should be used for converting math blocks/spans. If this option is set to +nil+, no math engine is used and the math blocks/spans are output as is.

Where does this have to be set? In `_config.yml`, in `Gemfile.lock` or somewhere else?

I noticed another strange problem which I believe has to do with this: When I write `$p(y|m)$` or similar in in-line math, Markdown/kramdown/whatever takes `|` as signalling a table, totally screwing up the math display. Currently, I have to replace `|` by `\mid` in order to solve this. Examples [here](https://statproofbook.github.io/Proofs/lme-der.html) and [here](https://statproofbook.github.io/Proofs/dmi-mce.html).

**Comment by allefeld** [comment]
> When I write $p(y|m)$ or similar in in-line math, Markdown/kramdown/whatever takes | as signalling a table
probably because `$` is not recognized as a math delimiter by kramdown. Still weird. Anyway, `\mid` is recommended for conditional probability because it acts as a binary operator, inserting space around it.

**Comment by allefeld** [comment]
> Where does this have to be set?

The page I linked to mentions three possibilities. But none of them mention a config file, so...

**Comment by StatProofBook** [comment]
> Anyway, `\mid` is recommended for conditional probability because it acts as a binary operator, inserting space around it.

Yes. I don't like the space. It's way too big.

**Comment by StatProofBook** [comment]
> Yes. I don't like the space. It's way too big.

I now found out that one can also use `\vert` (for conditional probabilities) or `\lvert` and `\rvert` (e.g. for determinants).

**Comment by StatProofBook** [comment]
For the record: I [changed the way](https://github.com/mathjax/MathJax/issues/2272#issuecomment-588216013) the MathJax fix is being called from `mathjax.html`, see 9bd6cac.

**Comment by JoramSoch** [comment]
Note to self: [dpvc's response](https://github.com/mathjax/MathJax/issues/2272#issuecomment-588287889) in MathJax issue #2272 discussion.
